In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('BCGX.csv')
df.head()

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Net Cash Flow from Operating
0,MSFT,2024,245122M,88136M,512163M,243686M,118548M
1,MSFT,2023,211915M,72361M,411976M,205753M,87582M
2,MSFT,2022,198270M,72738M,364840M,198298M,89035M
3,AAPL,2024,391035M,93736M,364980M,308030M,118254M
4,AAPL,2023,383285M,96995M,352583M,290437M,110543M


In [3]:
df.head()

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Net Cash Flow from Operating
0,MSFT,2024,245122M,88136M,512163M,243686M,118548M
1,MSFT,2023,211915M,72361M,411976M,205753M,87582M
2,MSFT,2022,198270M,72738M,364840M,198298M,89035M
3,AAPL,2024,391035M,93736M,364980M,308030M,118254M
4,AAPL,2023,383285M,96995M,352583M,290437M,110543M


In [4]:
df = df.sort_values(by=["Company", "Year"], ascending=[True, True]).reset_index(drop=True)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   Company                       9 non-null      object
 1   Year                          9 non-null      int64 
 2   Total Revenue                 9 non-null      object
 3   Net Income                    9 non-null      object
 4   Total Assets                  9 non-null      object
 5   Total Liabilities             9 non-null      object
 6   Net Cash Flow from Operating  9 non-null      object
dtypes: int64(1), object(6)
memory usage: 636.0+ bytes


In [6]:
# Function to convert M/B strings to billions
def convert_to_billions(x):
    if isinstance(x, str) and x.upper().endswith('M'):
        return float(x[:-1]) / 1000  # millions to billions

# Apply to all financial columns
cols_to_convert = ["Total Revenue", "Net Income", "Total Assets", "Total Liabilities", "Net Cash Flow from Operating"]
df[cols_to_convert] = df[cols_to_convert].applymap(convert_to_billions)

# Round for clarity
df[cols_to_convert] = df[cols_to_convert].round(2)

In [7]:
df

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Net Cash Flow from Operating
0,AAPL,2022,394.33,99.80,352.76,302.08,122.15
1,AAPL,2023,383.28,97.00,352.58,290.44,110.54
2,AAPL,2024,391.04,93.74,364.98,308.03,118.25
3,MSFT,2022,198.27,72.74,364.84,198.30,89.04
4,MSFT,2023,211.92,72.36,411.98,205.75,87.58
5,MSFT,2024,245.12,88.14,512.16,243.69,118.55
6,TSLA,2022,81.46,12.59,4.83,2.22,14.72
7,TSLA,2023,96.77,14.97,4.09,3.78,13.26
8,TSLA,2024,97.69,7.15,3.23,4.06,14.92


In [8]:
#YoY Trends
metrics = ["Total Revenue", "Net Income", "Total Assets", "Total Liabilities", "Net Cash Flow from Operating"]

for col in metrics:
    df[f"{col} YoY %"] = df.groupby("Company")[col].pct_change() * 100  # multiply by 100 for %
    df[f"{col} YoY %"] = df[f"{col} YoY %"].round(2)  # optional rounding

In [9]:
df.fillna(0, inplace=True)

In [10]:
df

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Net Cash Flow from Operating,Total Revenue YoY %,Net Income YoY %,Total Assets YoY %,Total Liabilities YoY %,Net Cash Flow from Operating YoY %
0,AAPL,2022,394.33,99.80,352.76,302.08,122.15,0.00,0.00,0.00,0.00,0.00
1,AAPL,2023,383.28,97.00,352.58,290.44,110.54,-2.80,-2.81,-0.05,-3.85,-9.50
2,AAPL,2024,391.04,93.74,364.98,308.03,118.25,2.02,-3.36,3.52,6.06,6.97
3,MSFT,2022,198.27,72.74,364.84,198.30,89.04,0.00,0.00,0.00,0.00,0.00
4,MSFT,2023,211.92,72.36,411.98,205.75,87.58,6.88,-0.52,12.92,3.76,-1.64
5,MSFT,2024,245.12,88.14,512.16,243.69,118.55,15.67,21.81,24.32,18.44,35.36
6,TSLA,2022,81.46,12.59,4.83,2.22,14.72,0.00,0.00,0.00,0.00,0.00
7,TSLA,2023,96.77,14.97,4.09,3.78,13.26,18.79,18.90,-15.32,70.27,-9.92
8,TSLA,2024,97.69,7.15,3.23,4.06,14.92,0.95,-52.24,-21.03,7.41,12.52


### 📊 Apple Financial Performance Summary (2022–2024)

- 2023 Dip:
Apple saw a decline in both revenue **(-2.8%)** and net income **(-2.8%)** in 2023. Operating cash flow dropped ~9.5%, indicating softer business performance and possible operational inefficiencies.
- 2024 Rebound (Partially):
Revenue recovered **(+2.0%)** and cash flow from operations rose by ~7%, showing signs of stabilization.
However, net income continued to decline **(-3.4%)**, suggesting profitability pressures — likely due to rising costs or margin erosion.
- Balance Sheet:
Assets grew modestly in 2024 **(+3.5%)**, while liabilities jumped 6%, indicating a possible uptick in borrowing or deferred expenses.

### 📊 Microsoft Financial Performance Summary (2022–2024)

- 2023: Stable Revenue, Flat Profits
Microsoft grew its revenue by **6.9%**, but net income slightly declined **(-0.5%)**, suggesting rising costs or flat margins. Cash flow from operations also dipped 1.6%, a minor operational signal.
- 2024: Strong Growth Across the Board
Revenue surged **15.7%**, and net income jumped **21.8%**, showing not just topline expansion but also improved profitability.
Operating cash flow soared **35.4%**, signaling efficient and profitable core operations.
- Balance Sheet Strength:
Assets grew **24.3%** in 2024, matched by an **18.4%** rise in liabilities.
Still, liabilities are less than 50% of revenue, indicating a healthy leverage position and good solvency.

### 📊Tesla Financial Performance Summary (2022–2024)

- 2023: Strong Revenue, Higher Debt
Revenue jumped **18.8%**, and net income grew **18.9%**, signaling good momentum.
However, assets shrank **(-15.3%)** while liabilities surged **(+70.3%)**, indicating Tesla may have taken on substantial new debt or restructured its balance sheet.
Operating cash flow dipped **(-9.9%)**, showing some stress in core operations.
- 2024: Profit Collapse, Slight Operational Recovery
Revenue stayed flat **(+0.95%)**, but net income fell sharply (-52.2%) — likely due to increased costs or lower margins.
Cash flow from operations improved **+12.5%**, hinting at internal efficiency gains, despite the profit drop.
Assets continued to decline **(-21%)**, while liabilities increased again **(+7.4%)**.

In [11]:
df['profit_margins']=df['Net Income']/df['Total Revenue'] * 100

In [12]:
df['profit_margins'] = df['profit_margins'].round(2)

In [13]:
df.head()

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Net Cash Flow from Operating,Total Revenue YoY %,Net Income YoY %,Total Assets YoY %,Total Liabilities YoY %,Net Cash Flow from Operating YoY %,profit_margins
0,AAPL,2022,394.33,99.80,352.76,302.08,122.15,0.00,0.00,0.00,0.00,0.00,25.31
1,AAPL,2023,383.28,97.00,352.58,290.44,110.54,-2.80,-2.81,-0.05,-3.85,-9.50,25.31
2,AAPL,2024,391.04,93.74,364.98,308.03,118.25,2.02,-3.36,3.52,6.06,6.97,23.97
3,MSFT,2022,198.27,72.74,364.84,198.30,89.04,0.00,0.00,0.00,0.00,0.00,36.69
4,MSFT,2023,211.92,72.36,411.98,205.75,87.58,6.88,-0.52,12.92,3.76,-1.64,34.14


In [14]:
df['risk']=df['Total Liabilities']/df['Total Assets'] # Indicated financial risk

In [15]:
df.head()

,Company,Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Net Cash Flow from Operating,Total Revenue YoY %,Net Income YoY %,Total Assets YoY %,Total Liabilities YoY %,Net Cash Flow from Operating YoY %,profit_margins,risk
0,AAPL,2022,394.33,99.80,352.76,302.08,122.15,0.00,0.00,0.00,0.00,0.00,25.31,0.856333
1,AAPL,2023,383.28,97.00,352.58,290.44,110.54,-2.80,-2.81,-0.05,-3.85,-9.50,25.31,0.823756
2,AAPL,2024,391.04,93.74,364.98,308.03,118.25,2.02,-3.36,3.52,6.06,6.97,23.97,0.843964
3,MSFT,2022,198.27,72.74,364.84,198.30,89.04,0.00,0.00,0.00,0.00,0.00,36.69,0.543526
4,MSFT,2023,211.92,72.36,411.98,205.75,87.58,6.88,-0.52,12.92,3.76,-1.64,34.14,0.499417


In [16]:
summary = df.groupby('Company').agg({
    'Total Revenue YoY %': 'mean',
    'Net Income YoY %': 'mean'
}).reset_index()

In [17]:
summary

,Company,Total Revenue YoY %,Net Income YoY %
0,AAPL,-0.260000,-2.056667
1,MSFT,7.516667,7.096667
2,TSLA,6.580000,-11.113333
